File: ios/DailyMemory/Domain/UseCase/AI/AnalyzeMemoryUseCase.swift
import Foundation

/// Use case for analyzing memory text with AI
/// Extracts entities like people, locations, dates, and amounts
final class AnalyzeMemoryUseCase {
    private let aiAnalysisService: AIAnalysisService

    init(aiAnalysisService: AIAnalysisService = .shared) {
        self.aiAnalysisService = aiAnalysisService
    }

    /// Analyze text and return extracted entities
    func execute(text: String) async -> Result<MemoryAnalysis, Error> {
        let result = await aiAnalysisService.analyzeText(text)

        switch result {
        case .success(let analysisResult):
            return .success(analysisResult.toMemoryAnalysis())
        case .failure(let error):
            return .failure(error)
        }
    }

    /// Analyze and create a Memory object with extracted data
    func analyzeAndCreateMemory(
        content: String,
        latitude: Double? = nil,
        longitude: Double? = nil
    ) async -> Result<Memory, Error> {
        let result = await aiAnalysisService.analyzeText(content)

        switch result {
        case .success(let analysisResult):
            let analysis = analysisResult.toMemoryAnalysis()

            let memory = Memory(
                content: content,
                extractedPersons: analysis.persons,
                extractedLocation: analysis.location,
                extractedDate: analysis.date,
                extractedAmount: analysis.amount,
                extractedTags: analysis.tags,
                category: analysis.category,
                recordedAt: Date(),
                recordedLatitude: latitude,
                recordedLongitude: longitude
            )

            return .success(memory)

        case .failure(let error):
            return .failure(error)
        }
    }
}

/// Data class for memory analysis results
struct MemoryAnalysis {
    let persons: [String]
    let location: String?
    let date: Date?
    let amount: Double?
    let tags: [String]
    let category: Category
    let summary: String
}

/// Extension to convert API result to domain model
extension AnalysisResult {
    func toMemoryAnalysis() -> MemoryAnalysis {
        // Parse date if present
        var parsedDate: Date?
        if let dateString = date {
            let formatter = ISO8601DateFormatter()
            formatter.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
            parsedDate = formatter.date(from: dateString)

            // Try without fractional seconds
            if parsedDate == nil {
                formatter.formatOptions = [.withInternetDateTime]
                parsedDate = formatter.date(from: dateString)
            }

            // Try date only
            if parsedDate == nil {
                formatter.formatOptions = [.withFullDate]
                parsedDate = formatter.date(from: dateString)
            }
        }

        // Parse category
        let parsedCategory = Category(rawValue: category.uppercased()) ?? .general

        return MemoryAnalysis(
            persons: persons,
            location: location,
            date: parsedDate,
            amount: amount,
            tags: tags,
            category: parsedCategory,
            summary: summary
        )
    }
}

File: android/app/src/main/java/com/dailymemory/domain/usecase/ai/AnalyzeMemoryUseCase.kt
package com.dailymemory.domain.usecase.ai

import com.dailymemory.data.remote.AIAnalysisService
import com.dailymemory.data.remote.AnalysisResult
import com.dailymemory.domain.model.Category
import com.dailymemory.domain.model.Memory
import java.time.LocalDateTime
import java.time.format.DateTimeFormatter
import javax.inject.Inject

/**
 * Use case for analyzing memory text with AI
 * Extracts entities like people, locations, dates, and amounts
 */
class AnalyzeMemoryUseCase @Inject constructor(
    private val aiAnalysisService: AIAnalysisService
) {
    /**
     * Analyze text and return extracted entities
     */
    suspend operator fun invoke(text: String): Result<MemoryAnalysis> {
        return aiAnalysisService.analyzeText(text).map { result ->
            result.toMemoryAnalysis()
        }
    }

    /**
     * Analyze and create a Memory object with extracted data
     */
    suspend fun analyzeAndCreateMemory(
        content: String,
        latitude: Double? = null,
        longitude: Double? = null
    ): Result<Memory> {
        return aiAnalysisService.analyzeText(content).map { result ->
            val analysis = result.toMemoryAnalysis()

            Memory(
                content = content,
                extractedPersons = analysis.persons,
                extractedLocation = analysis.location,
                extractedDate = analysis.date,
                extractedAmount = analysis.amount,
                extractedTags = analysis.tags,
                category = analysis.category,
                recordedAt = LocalDateTime.now(),
                recordedLatitude = latitude,
                recordedLongitude = longitude
            )
        }
    }
}

/**
 * Data class for memory analysis results
 */
data class MemoryAnalysis(
    val persons: List<String>,
    val location: String?,
    val date: LocalDateTime?,
    val amount: Double?,
    val tags: List<String>,
    val category: Category,
    val summary: String
)

/**
 * Extension to convert API result to domain model
 */
private fun AnalysisResult.toMemoryAnalysis(): MemoryAnalysis {
    // Parse date if present
    val parsedDate = date?.let {
        try {
            LocalDateTime.parse(it, DateTimeFormatter.ISO_DATE_TIME)
        } catch (e: Exception) {
            try {
                LocalDateTime.parse(it + "T00:00:00", DateTimeFormatter.ISO_DATE_TIME)
            } catch (e: Exception) {
                null
            }
        }
    }

    // Parse category
    val parsedCategory = try {
        Category.valueOf(category.uppercase())
    } catch (e: Exception) {
        Category.GENERAL
    }

    return MemoryAnalysis(
        persons = persons,
        location = location,
        date = parsedDate,
        amount = amount,
        tags = tags,
        category = parsedCategory,
        summary = summary
    )
}
